File: typescript/kata1/src/ShowScheduler/ShowScheduler.ts
import { Show } from "../Show/Show.js";

export class ShowScheduler {
  private shows: Show[];

  constructor(shows: Show[]) {
    this.shows = shows;
  }

  public calculateTotalDuration(): number {
    let totalDuration = 0;

    for (const show of this.shows) {
      totalDuration += this.calculateShowDuration(show);
    }

    return totalDuration;
  }

  public getShowDurationByName(name: string): number {
    const show = this.shows.find((show) => show.name === name);

    if (!show) {
      throw new Error("Show isn't scheduled");
    }

    return this.calculateShowDuration(show);
  }

  private calculateShowDuration(show: Show): number {
    let showDuration: number;

    if (show.type === "clown") {
      showDuration = show.jokes!.length * 5;
    } else if (show.type === "tightrope") {
      showDuration = show.ropeLength! / 2 + (show.isBlindfolded ? 5 : 0);
    } else if (show.type === "trapeze") {
      showDuration = show.performers! * 6;
    } else {
      showDuration = 0;
    }

    return showDuration;
  }
}

File: java/kata1/src/main/java/com/therefactorproject/showscheduler/ShowScheduler.java
package com.therefactorproject.showscheduler;

import com.therefactorproject.show.Show;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

import static java.util.stream.Collectors.toMap;

public class ShowScheduler {
  private final Map<String, Show> showsByName;

  public ShowScheduler(List<Show> shows) {
    this.showsByName = shows.stream()
      .collect(toMap(Show::name, Function.identity()));
  }

  public double calculateTotalDuration() {
    var totalDuration = 0.0;

    for (var show: showsByName.values()) {
      totalDuration += show.duration();
    }

    return totalDuration;
  }

  public double getShowDurationByName(String name) {
    return Optional.ofNullable(showsByName.get(name))
      .map(Show::duration)
      .orElseThrow(() -> new IllegalArgumentException("Show isn't scheduled"));
  }
}
